File: clojure/crypto-square/src/crypto_square.clj
(ns crypto-square
  (:require [clojure.string :as str]))

(defn normalize-plaintext [plaintext] (str/lower-case (str/replace plaintext #"[^\w]" "")))

(defn square-size [text] (->> text
                              (count)
                              (Math/sqrt)
                              (Math/ceil)
                              (int)))

(defn plaintext-segments [plaintext]
  (let [text (normalize-plaintext plaintext)
        length (square-size text)]
    (map str/join (partition length length "" (str/split text #"")))))

(defn- pad-to [s length] (format (str "%-" length "s") s))

(defn equalise-segment-lengths [segments required-length]
  (map #(pad-to % required-length) segments))

(defn ciphertext [plaintext]
  (let [column-length (square-size (normalize-plaintext plaintext))
        segments (plaintext-segments plaintext)
        segments-eq (equalise-segment-lengths segments column-length)
        columns (apply interleave segments-eq)]
    (str/replace (str/join columns) #"[ ]+" "")))

(defn- format-ciphertext [ciphertext]
  (->> ciphertext
       (map #(str/trim (str/join %)))
       (str/join " ")))

(defn normalize-ciphertext [plaintext]
  (let [row-length (count (plaintext-segments plaintext))
        column-length (square-size (normalize-plaintext plaintext))
        ciphertext (ciphertext plaintext)
        maximum-length (* row-length column-length)
        length-shortfall (- maximum-length (count ciphertext))
        full-segments-length (* row-length (- column-length length-shortfall))
        full-segments (->>  ciphertext
                            (take full-segments-length)
                            (partition row-length))
        short-segments (->> ciphertext
                            (drop full-segments-length)
                            (partition (dec row-length)))]
    (format-ciphertext (concat full-segments short-segments))))

File: kotlin/crypto-square/src/main/kotlin/CryptoSquare.kt
import kotlin.math.ceil
import kotlin.math.sqrt

object CryptoSquare {

    fun ciphertext(plaintext: String): String {
        val chars = plaintext.filter{ it.isLetterOrDigit() }.lowercase()

        if (chars.isEmpty()) {
            return ""
        }

        val numberOfColumns = ceil(sqrt(chars.length.toDouble())).toInt()

        val rows = chars.chunked(numberOfColumns) { it.padEnd(numberOfColumns, ' ') }

        var columns = mutableListOf<String>()
        for (i in 0 until numberOfColumns) {
            val column = rows.map { it[i] }
            columns.add(column.joinToString(""))
        }

        return columns.joinToString(" ")
    }

}
